autoprecompiles/src/constraint_optimizer.rs

```diff
@@ -63,14 +63,14 @@ fn solver_based_optimization<T: FieldElement, V: Clone + Ord + Hash + Display>(
     mut constraint_system: JournalingConstraintSystem<T, V>,
     bus_interaction_handler: impl BusInteractionHandler<T>,
 ) -> Result<JournalingConstraintSystem<T, V>, Error> {
-    let result = Solver::new(constraint_system.system().clone())
-        .with_bus_interaction_handler(bus_interaction_handler)
-        .solve()?;
+    let result = solve_system(constraint_system.system().clone(), bus_interaction_handler)?;
     log::trace!("Solver figured out the following assignments:");
     for (var, value) in result.assignments.iter() {
         log::trace!("  {var} = {value}");
     }
     constraint_system.apply_substitutions(result.assignments);
+    // TODO could we somehow get rid of this special case by keeping
```

Comment:
Also relevant: https://github.com/powdr-labs/powdr/issues/2980

Reply:
Hmm, it seems like introducing these variables is kind of a solver implementation detail. It's kind of a hack...

And I think if we would keep them longer, the inliner would just inline them?